src/info/langs/language.tera

```diff
@@ -109,6 +110,34 @@ impl Language {
     }
 }
 
+/// Counts the lines-of-code of a tokei `Language`. Takes into
+/// account that a prose language's comments *are* its code.
+pub fn loc(language_type: &tokei::LanguageType, language: &tokei::Language) -> usize {
+    let loc = match language_type {
+        {% for language, attrs in languages -%}
+            {%- set line_types = attrs.line_types | default(value=['code']) -%}
+            tokei::LanguageType::{{ language }} => language.{{ line_types.0 }}{% for line_type in line_types | slice(start=1) %} + language.{{ line_type }}{% endfor %},
+        {% endfor %}
+        _ => unimplemented!("Language Type {:?}", language_type),
+    };
+    loc + language.children.iter().fold(0, |sum, (lang_type, reports)| {
+        sum + reports.iter().fold(0, |sum, report| sum + stats_loc(lang_type, &report.stats))
+    })
+}
+
+/// Counts the lines-of-code of a tokei `Report`. This is the child of a
+/// `tokei::CodeStats`.
+pub fn stats_loc(language_type: &tokei::LanguageType, stats: &tokei::CodeStats) -> usize {
+    let stats = stats.summarise();
+    match language_type {
+        {% for language, attrs in languages -%}
+            {%- set line_types = attrs.line_types | default(value=['code']) -%}
+            tokei::LanguageType::{{ language }} => stats.{{ line_types.0 }}{% for line_type in line_types | slice(start=1) %} + stats.{{ line_type }}{% endfor %},
+        {% endfor %}
+        _ => unimplemented!("Language Type {:?}", language_type),
+    }
+}
```

Comment:
Don't you think it would be better to keep the code logic outside of the tera template? I'm refering to lines 123-125 and 131. 

We could probably extract those lines and put them in `get_language_distribution()` between line 34 and 36:

```
loc += language
  .children
  .iter()
  .fold(0, |sum, (lang_type, reports)| {
      sum + reports.iter().fold(0, |sum, report| {
          sum + language::stats_loc(lang_type, &report.stats.summarise())
      })
  });
```

this would simplify the template but more importantly avoid obfuscating part of the code logic.

In the template, we would then have two methods `loc` and `stats_loc` which would be almost identical except for one of their parameters. I am wonder if there is a way we could combine the two methods 🤔 (maybe in the future)

Reply:
:thinking: Yeah, it does obfuscate it a bit by hiding the logic in the template. But my thinking was that I'd like to avoid needing to repeat similar logic (adding the LOC of the children) for both language distributions and total line counts. Since these are kind of private methods, maybe they should be `#[doc(hidden)]` and called `__loc()`, and have more visible methods called `loc()` and `stats_loc()` in `language.rs` that call these methods and act as wrappers.

```rust
fn loc(language_type: &tokei::LanguageType, language: &tokei::Language) -> usize {
    __loc(language_type, language) + language
        .children 
        // ...
```
What do you think?

Reply:
As you mentioned in your comment https://github.com/o2sh/onefetch/pull/937#issuecomment-1399553995
We'll have to add some unit tests of more complex cases (language with children) to make sure we don't break things.

Reply:
> But my thinking was that I'd like to avoid needing to repeat similar logic (adding the LOC of the children) for both language distributions and total line counts.

You're right, I missed that 🤦 

> Since these are kind of private methods, maybe they should be #[doc(hidden)] and called __loc(), and have more visible methods called loc() and stats_loc() in language.rs that call these methods and act as wrappers.

I'm sorry, but I'm lost - too many `loc` 😅 - maybe you can commit your suggestion to illustrate your idea. But yeah, if we could avoid putting logic inside the tera template, that would be great. we could probably factorize this part inside a method:

```
let mut loc = language::loc(language_name, language);

let has_children = !language.children.is_empty();

if has_children {
    for (lang_type, reports) in language.children {
        loc += reports
            .iter()
            .map(|r| r.stats.summarise())
            .fold(0, |sum, stats| sum + language::stats_loc(&lang_type, &stats));
    }
}
```

and call it for both language distributions and total line counts.

But it's your call, do what you think is best 👍 I think the PR is already pretty good.

Reply:
> maybe you can commit your suggestion to illustrate your idea.

Sure! https://github.com/o2sh/onefetch/pull/937/commits/3e14718ac30022257e837fbc1e36bd24b91962df is the basic idea is that `__<function>` contains the repetitive code that the template manages, and `<function>` is a wrapper that calls that function and includes some helper code.

Reply:
I'll probably add a test for Jupyter code and call it on `get_total_loc`. That way `loc` and `stats_loc` is 100% covered. `__loc`'s *template* would be 100% covered, but the *generated code* would probably be <10% with each language being a branch :laughing:
I think I'll just hardcode the structs necessary to simulate Jupyter code.

Reply:
Perfect, thanks @spenserblack